File: capture-tui-go/internal/models/entry.go
package models

import (
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/google/uuid"
	"gopkg.in/yaml.v3"
)

// Entry represents a captured idea or task
type Entry struct {
	ID        string         `json:"id" yaml:"id"`
	Title     string         `json:"title" yaml:"title"`
	Content   string         `json:"content" yaml:"content"`
	Category  string         `json:"category" yaml:"category"`
	Tags      []string       `json:"tags" yaml:"tags"`
	CreatedAt time.Time      `json:"created_at" yaml:"created_at"`
	UpdatedAt *time.Time     `json:"updated_at,omitempty" yaml:"updated_at,omitempty"`
	Tasks     []string       `json:"tasks" yaml:"tasks"`
	Metadata  EntryMetadata  `json:"metadata" yaml:"metadata"`
}

// EntryMetadata contains additional metadata for an entry
type EntryMetadata struct {
	Source         string   `json:"source" yaml:"source"`
	Priority       string   `json:"priority" yaml:"priority"`
	Author         string   `json:"author,omitempty" yaml:"author,omitempty"`
	RelatedEntries []string `json:"related_entries,omitempty" yaml:"related_entries,omitempty"`
}

// NewEntry creates a new entry with generated ID
func NewEntry(title, content, category string, tags []string) *Entry {
	now := time.Now()
	return &Entry{
		ID:        generateID(),
		Title:     title,
		Content:   content,
		Category:  category,
		Tags:      tags,
		CreatedAt: now,
		Metadata: EntryMetadata{
			Source:   "cli",
			Priority: "P2",
		},
	}
}

// generateID generates a unique entry ID
func generateID() string {
	timestamp := time.Now().Format("20060102-150405")
	shortUUID := uuid.New().String()[:6]
	return fmt.Sprintf("ideas-%s-%s", timestamp, shortUUID)
}

// ToMarkdown converts the entry to Markdown format
func (e *Entry) ToMarkdown() string {
	var sb strings.Builder

	// Frontmatter
	sb.WriteString("---\n")
	sb.WriteString(fmt.Sprintf("id: \"%s\"\n", e.ID))
	sb.WriteString(fmt.Sprintf("category: \"%s\"\n", e.Category))
	sb.WriteString(fmt.Sprintf("created_at: \"%s\"\n", e.CreatedAt.Format(time.RFC3339)))
	sb.WriteString(fmt.Sprintf("tags: %v\n", e.Tags))
	sb.WriteString(fmt.Sprintf("priority: \"%s\"\n", e.Metadata.Priority))
	sb.WriteString(fmt.Sprintf("source: \"%s\"\n", e.Metadata.Source))
	sb.WriteString("---\n\n")

	// Title
	sb.WriteString(fmt.Sprintf("# %s\n\n", e.Title))

	// Content
	sb.WriteString(e.Content)
	sb.WriteString("\n")

	// Tasks
	if len(e.Tasks) > 0 {
		sb.WriteString("\n## 关联任务\n\n")
		for _, task := range e.Tasks {
			sb.WriteString(fmt.Sprintf("- [ ] %s\n", task))
		}
	}

	return sb.String()
}

// EntryFromMarkdown parses an entry from Markdown content
func EntryFromMarkdown(content string) (*Entry, error) {
	// Extract frontmatter
	frontmatterRegex := regexp.MustCompile(`(?s)^---\n(.*?)\n---\n(.*)`)
	matches := frontmatterRegex.FindStringSubmatch(content)
	if len(matches) < 3 {
		return nil, fmt.Errorf("invalid markdown format: missing frontmatter")
	}

	var frontmatter map[string]interface{}
	if err := yaml.Unmarshal([]byte(matches[1]), &frontmatter); err != nil {
		return nil, fmt.Errorf("failed to parse frontmatter: %w", err)
	}

	body := matches[2]

	// Extract title
	titleRegex := regexp.MustCompile(`(?m)^# (.+)$`)
	titleMatch := titleRegex.FindStringSubmatch(body)
	title := "Untitled"
	if len(titleMatch) > 1 {
		title = strings.TrimSpace(titleMatch[1])
	}

	// Extract content (remove title)
	contentBody := titleRegex.ReplaceAllString(body, "")
	contentBody = strings.TrimSpace(contentBody)

	// Extract tasks
	taskRegex := regexp.MustCompile(`(?m)^- \[ \] (.+)$`)
	taskMatches := taskRegex.FindAllStringSubmatch(contentBody, -1)
	tasks := make([]string, 0, len(taskMatches))
	for _, match := range taskMatches {
		if len(match) > 1 {
			tasks = append(tasks, strings.TrimSpace(match[1]))
		}
	}

	// Remove tasks section from content
	tasksSectionRegex := regexp.MustCompile(`(?s)\n?## 关联任务\n\n(?:- \[ \] .+\n?)+`)
	contentBody = tasksSectionRegex.ReplaceAllString(contentBody, "")
	contentBody = strings.TrimSpace(contentBody)

	// Parse timestamps
	createdAt, _ := time.Parse(time.RFC3339, getString(frontmatter, "created_at"))

	entry := &Entry{
		ID:        getString(frontmatter, "id"),
		Title:     title,
		Content:   contentBody,
		Category:  getString(frontmatter, "category"),
		Tags:      getStringSlice(frontmatter, "tags"),
		CreatedAt: createdAt,
		Tasks:     tasks,
		Metadata: EntryMetadata{
			Source:   getString(frontmatter, "source"),
			Priority: getString(frontmatter, "priority"),
		},
	}

	return entry, nil
}

// Helper functions
func getString(m map[string]interface{}, key string) string {
	if v, ok := m[key]; ok {
		if s, ok := v.(string); ok {
			return s
		}
	}
	return ""
}

func getStringSlice(m map[string]interface{}, key string) []string {
	if v, ok := m[key]; ok {
		if arr, ok := v.([]interface{}); ok {
			result := make([]string, 0, len(arr))
			for _, item := range arr {
				if s, ok := item.(string); ok {
					result = append(result, s)
				}
			}
			return result
		}
	}
	return []string{}
}

// Update updates the entry's updated_at timestamp
func (e *Entry) Update() {
	now := time.Now()
	e.UpdatedAt = &now
}

// AddTask adds a task to the entry
func (e *Entry) AddTask(task string) {
	e.Tasks = append(e.Tasks, task)
}

// HasTasks returns true if the entry has tasks
func (e *Entry) HasTasks() bool {
	return len(e.Tasks) > 0
}

// TaskCount returns the number of tasks
func (e *Entry) TaskCount() int {
	return len(e.Tasks)
}

File: capture/capture_tui/models/entry.py
"""Entry 数据模型"""

from dataclasses import dataclass, field, asdict
from datetime import datetime
from typing import List, Optional
import uuid


@dataclass
class EntryMetadata:
    """条目元数据"""
    source: str = "cli"  # cli, file, session
    priority: str = "P2"  # P0, P1, P2
    author: Optional[str] = None
    related_entries: List[str] = field(default_factory=list)
    
    def to_dict(self) -> dict:
        return asdict(self)
    
    @classmethod
    def from_dict(cls, data: dict) -> "EntryMetadata":
        return cls(**data)


@dataclass
class Entry:
    """想法/任务条目"""
    id: str
    title: str
    content: str
    category: str
    tags: List[str] = field(default_factory=list)
    created_at: datetime = field(default_factory=datetime.now)
    updated_at: Optional[datetime] = None
    tasks: List[str] = field(default_factory=list)
    metadata: EntryMetadata = field(default_factory=EntryMetadata)
    
    def __post_init__(self):
        if self.updated_at is None:
            self.updated_at = self.created_at
    
    @classmethod
    def create(
        cls,
        title: str,
        content: str,
        category: str = "uncategorized",
        tags: Optional[List[str]] = None,
        tasks: Optional[List[str]] = None,
        metadata: Optional[EntryMetadata] = None
    ) -> "Entry":
        """创建新条目"""
        entry_id = cls._generate_id()
        return cls(
            id=entry_id,
            title=title,
            content=content,
            category=category,
            tags=tags or [],
            tasks=tasks or [],
            metadata=metadata or EntryMetadata()
        )
    
    @staticmethod
    def _generate_id() -> str:
        """生成唯一ID"""
        timestamp = datetime.now().strftime("%Y%m%d-%H%M%S")
        short_uuid = uuid.uuid4().hex[:6]
        return f"ideas-{timestamp}-{short_uuid}"
    
    def to_dict(self) -> dict:
        """转换为字典"""
        return {
            "id": self.id,
            "title": self.title,
            "content": self.content,
            "category": self.category,
            "tags": self.tags,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
            "tasks": self.tasks,
            "metadata": self.metadata.to_dict()
        }
    
    @classmethod
    def from_dict(cls, data: dict) -> "Entry":
        """从字典创建"""
        return cls(
            id=data["id"],
            title=data["title"],
            content=data["content"],
            category=data["category"],
            tags=data.get("tags", []),
            created_at=datetime.fromisoformat(data["created_at"]),
            updated_at=datetime.fromisoformat(data["updated_at"]) if data.get("updated_at") else None,
            tasks=data.get("tasks", []),
            metadata=EntryMetadata.from_dict(data.get("metadata", {}))
        )
    
    def to_markdown(self) -> str:
        """转换为 Markdown 格式"""
        lines = [
            "---",
            f'id: "{self.id}"',
            f'category: "{self.category}"',
            f'created_at: "{self.created_at.isoformat()}"',
            f'tags: {self.tags}',
            f'priority: "{self.metadata.priority}"',
            f'source: "{self.metadata.source}"',
            "---",
            "",
            f"# {self.title}",
            "",
            self.content,
            "",
        ]
        
        if self.tasks:
            lines.extend([
                "## 关联任务",
                ""
            ])
            for task in self.tasks:
                lines.append(f"- [ ] {task}")
            lines.append("")
        
        return "\n".join(lines)
    
    @classmethod
    def from_markdown(cls, content: str) -> "Entry":
        """从 Markdown 解析"""
        import re
        
        # 解析 frontmatter
        frontmatter_match = re.match(r'^---\n(.*?)\n---\n(.*)', content, re.DOTALL)
        if not frontmatter_match:
            raise ValueError("Invalid markdown format: missing frontmatter")
        
        import yaml
        fm = yaml.safe_load(frontmatter_match.group(1))
        body = frontmatter_match.group(2).strip()
        
        # 提取标题
        title_match = re.match(r'^# (.+)$', body, re.MULTILINE)
        title = title_match.group(1) if title_match else "Untitled"
        
        # 提取内容（去掉标题）
        content_body = re.sub(r'^# .+\n', '', body).strip()
        
        # 分离关联任务部分
        task_section_match = re.search(r'## 关联任务\n\n((?:- \[ \] .+\n?)+)', content_body)
        if task_section_match:
            # 从任务部分提取任务
            tasks = re.findall(r'- \[ \] (.+)', task_section_match.group(1))
            # 从内容中移除任务部分
            content_body = re.sub(r'\n?## 关联任务\n\n(?:- \[ \] .+\n?)+', '', content_body).strip()
        else:
            # 没有单独的任务部分，从整个内容提取
            tasks = re.findall(r'- \[ \] (.+)', content_body)
        
        metadata = EntryMetadata(
            source=fm.get("source", "file"),
            priority=fm.get("priority", "P2")
        )
        
        return cls(
            id=fm["id"],
            title=title,
            content=content_body,
            category=fm["category"],
            tags=fm.get("tags", []),
            created_at=datetime.fromisoformat(fm["created_at"]),
            metadata=metadata,
            tasks=tasks
        )
